controlplane/kubeadm/controllers/controller.go

```diff
@@ -57,7 +57,7 @@ import (
 // +kubebuilder:rbac:groups=infrastructure.cluster.x-k8s.io;bootstrap.cluster.x-k8s.io;controlplane.cluster.x-k8s.io,resources=*,verbs=get;list;watch;create;update;patch;delete
 // +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=clusters;clusters/status,verbs=get;list;watch
 // +kubebuilder:rbac:groups=cluster.x-k8s.io,resources=machines;machines/status,verbs=get;list;watch;create;update;patch;delete
-// +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list
+// +kubebuilder:rbac:groups=apiextensions.k8s.io,resources=customresourcedefinitions,verbs=get;list;watch
```

Comment:
I think we have to figure out why/how we're watching CRDs and stop doing that instead

Reply:
It looks like the following is creating the watch

https://github.com/kubernetes-sigs/cluster-api/blob/fc52fe3d5cfe05d436a7d2f0ba3cf8a1399aad59/controlplane/kubeadm/controllers/helpers.go#L129

https://github.com/kubernetes-sigs/cluster-api/blob/ef22a29d4aa71e975d509a098d698c0a3ed09f9e/util/conversion/conversion.go#L65

https://github.com/kubernetes-sigs/cluster-api/blob/556258b232881d2167010b35c8200a51c87c15bb/util/util.go#L426

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/split.go#L124-L131
(we're doing a Get with a PartialObjectMeta, but this block is checking if CRDs should bypass the cache)

https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/cache/informer_cache.go#L54
(CacheReader is informerCache, this step will create the informer and thus the watch)

Reply:
> It looks like the following is creating the watch
> 
> https://github.com/kubernetes-sigs/cluster-api/blob/fc52fe3d5cfe05d436a7d2f0ba3cf8a1399aad59/controlplane/kubeadm/controllers/helpers.go#L129
> 

How should we go about this line then?

Reply:
Summary:
* we're doing a get with PartialObjectMeta to get the the metadata of a CRD
* controller-runtime checks if CRDs should be cached. They are currently not on our uncached list so the answer is yes
* controller-runtime creates an informer and gets the whole CRD (not a PartialObjectMeta) from the server but returns the correct PartialObjectMeta

We can fix this on our side by disabling the cache for CRDs:
```go
		ClientDisableCacheFor: []client.Object{
			&corev1.ConfigMap{},
			&corev1.Secret{},
			&apiextensionsv1.CustomResourceDefinition{},
		},
```

But for me this looks like a controller-runtime bug.

@vincepri WDYT?

Reply:
@kashifest Sorry didn't saw your response while I was writing :). I think feedback from Vince would be good to figure out if the behavior of CR is intended. 

We might also regard the watch as desired behavior and then your fix is perfectly fine. If it isn't we might also have to take a look at other controllers.

Reply:
> I think we have to figure out why/how we're watching CRDs and stop doing that instead

Just for my knowledge, whats the harm if we add the watch here?

Reply:
And it would be also good to know currently because of the bug, how is it impacting the KCP workflow?

Reply:
The problem is that we're listing and watching all CRDs of the cluster. Depending on how much CRDs there are in the cluster this leads to significantly more RAM usage of the controller (I saw CRDs with kb's of schema).

Also the idea of using PartialObjectMeta was to be as efficient as possible and only get the metadata of the one CRD we want to take a look at. Listing and watching on all CRDs defeats that purpose.

Reply:
> The problem is that we're listing and watching all CRDs of the cluster. Depending on how much CRDs there are in the cluster this leads to significantly more RAM usage of the controller (I saw CRDs with kb's of schema).
> 
> Also the idea of using PartialObjectMeta was to be as efficient as possible and only get the metadata of the one CRD we want to take a look at. Listing and watching on all CRDs defeats that purpose.

Aha, thanks. I wasn't aware of it.

Reply:
We're watching CRDs here, although they should all be using PartialObjectMetadata, which should be ok-ish?

Reply:
> I'm wondering though, `log.Info("Cannot retrieve CRD with metadata only client, falling back to slower listing", "err", err.Error())` this log line should have showed up at some point as well, do we know what the error was?


The error was `E1116 07:57:06.965169       1 reflector.go:138] pkg/mod/k8s.io/client-go@v0.22.2/tools/cache/reflector.go:167: Failed to watch *v1.PartialObjectMetadata: unknown` . I didn't see the error in the log you mentioned. Perhaps it didn't reach that far?

Reply:
The partialObjectMetadata get will still use watch on CRD's using partial objects meta, which doesn't consume lots of resources and it's probably fine to keep — @sbueringer could we open a PR to disable the fallback listing?

Reply:
Yup will do that at the latest tomorrow.

@kashifest tl;dr is we can go ahead with this PR as is :)

Reply:
@vincepri cherry-pick to release-0.4 & release-1.0? (release-0.3 is not affected)

Reply:
> @vincepri cherry-pick to release-0.4 & release-1.0? (release-0.3 is not affected)

Is there any cherry-pick bot? Otherwise I can open prs towards the  branches as well

Reply:
We have a bot, can be triggered via:
```bash
/cherrypick release-1.0
/cherrypick release-0.4
```

Reply:
@vincepri I've created a issue for now: https://github.com/kubernetes-sigs/cluster-api/issues/5686#issue-1055367559